Source/RapyutaSimulationPlugins/Private/Core/RRCoreUtils.cpp

```diff
@@ -478,3 +480,173 @@ bool URRCoreUtils::LoadIESProfilesFromFolder(const FString& InFolderPath,
 
     return bResult;
 }
+
+// -------------------------------------------------------------------------------------------------------------------------
+// GRAPHICS UTILS --
+//
+#if WITH_EDITOR
+bool URRCoreUtils::RenderThumbnail(UObject* InObject,
+                                   uint32 InImageWidth,
+                                   uint32 InImageHeight,
+                                   const ThumbnailTools::EThumbnailTextureFlushMode::Type InFlushMode,
+                                   FObjectThumbnail* OutThumbnail)
+{
+    // Renderer must be initialized before generating thumbnails
+    if (!ensure(FApp::CanEverRender()))
+    {
+        return false;
+    }
+    if (!ensure(GIsRHIInitialized))
+    {
+        return false;
+    }
+
+    // Get the rendering info for this object
+    FThumbnailRenderingInfo* renderInfo = UThumbnailManager::Get().GetRenderingInfo(InObject);
+    if ((nullptr == renderInfo) || (nullptr == renderInfo->Renderer))
+    {
+        UE_LOG_WITH_INFO_SHORT(LogRapyutaCore,
+                               Error,
+                               TEXT("[ThumbnailManager] No rendering info found for class [%s]"),
+                               *InObject->GetClass()->GetName());
+        return false;
+    }
+
+    if (false == renderInfo->Renderer->CanVisualizeAsset(InObject))
+    {
+        UE_LOG_WITH_INFO_SHORT(LogRapyutaCore,
+                               Error,
+                               TEXT("[%s] cannot be visualized, probably due to invalid rendering resource"),
+                               *InObject->GetName());
+        return false;
+    }
+
+    // Thumbnail size
+    if (OutThumbnail)
+    {
+        OutThumbnail->SetImageSize(InImageWidth, InImageHeight);
+    }
+
+    // Create a static texture render target, as commonly used for all thumbnails
+    static UTextureRenderTarget2D* sThumbnailRenderTarget = []()
+    {
+        // Create thumbnail render target
+        UTextureRenderTarget2D* renderTarget = NewObject<UTextureRenderTarget2D>(GetTransientPackage());
+        renderTarget->AddToRoot();
+        renderTarget->ClearColor = FLinearColor::Transparent;
+        renderTarget->TargetGamma = GEngine->GetDisplayGamma();
+        renderTarget->RenderTargetFormat = RTF_RGBA8;
+        return renderTarget;
+    }();
+    sThumbnailRenderTarget->InitCustomFormat(InImageWidth, InImageHeight, PF_B8G8R8A8, false);
```

Comment:
The pixel format here does not match the render target format a few lines above. I suppose this was tested and works okay, but you might check that there isn't an RTF_BGRA8 for the above line. (For what it's worth, in modern graphics APIs, BRGA is the only guaranteed-to-be-supported format for render targets.)

Reply:
Right, it looks like `ETextureRenderTargetFormat` is UE-defined only for render target, and there is a specific mapping to the pixel format here
https://github.com/EpicGames/UnrealEngine/blob/release/Engine/Source/Runtime/Engine/Classes/Engine/TextureRenderTarget2D.h#L52
So the latter setting is redundant, will remove it.
Also, I take refs from here
https://github.com/EpicGames/UnrealEngine/blob/release/Engine/Source/Editor/UnrealEd/Private/ObjectTools.cpp#L5057
https://github.com/EpicGames/UnrealEngine/blob/release/Engine/Source/Editor/UnrealEd/Private/EditorEngine.cpp#L2766